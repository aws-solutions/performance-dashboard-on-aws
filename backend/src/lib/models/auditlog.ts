/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum ItemEvent {
    Create = "Create",
    Update = "Update",
    Delete = "Delete",
}

export type ModifiedProperty = {
    property: string;
    oldValue: any;
    newValue: any;
};

export interface AuditLog {
    itemId: string;
    timestamp: Date;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

// Represents a dynamodb item in the Audit Trail table
export interface AuditLogItem {
    pk: string;
    sk: string;
    type: string;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

export interface DashboardAuditLog extends AuditLog {
    version: number;
}

// Represents a dynamodb item in the Audit Trail table that
// contains attributes specific to Dashboard audit logs.
export interface DashboardAuditLogItem extends AuditLogItem {
    version: number;
}
