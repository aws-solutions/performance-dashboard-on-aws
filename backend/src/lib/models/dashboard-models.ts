export type Dashboard = {
    id: string,
    name: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};

export type DashboardList = Array<Dashboard>;

export interface DashboardItem {
    pk: string,
    sk: string,
    type: string,
    dashboardName: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};