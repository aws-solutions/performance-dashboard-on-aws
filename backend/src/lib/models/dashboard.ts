/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import { Widget } from "./widget";

export const DASHBOARD_ITEM_TYPE = "Dashboard";

export enum DashboardState {
    Draft = "Draft",
    Published = "Published",
    Archived = "Archived",
    /**
     * @type {string}
     * @deprecated in favor of {@link Draft}
     */
    PublishPending = "PublishPending",
    Inactive = "Inactive",
}

export interface DashboardVersion {
    id: string;
    version: number;
    state: DashboardState;
    friendlyURL?: string;
}

export interface Dashboard {
    id: string;
    name: string;
    version: number;
    parentDashboardId: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description: string;
    createdBy: string;
    updatedAt: Date;
    updatedBy?: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    deletedBy?: string;
    state: DashboardState;
    releaseNotes?: string;
    widgets?: Array<Widget>;
    friendlyURL?: string;
}

export type DashboardList = Array<Dashboard>;

export interface DashboardItem {
    pk: string;
    sk: string;
    type: string;
    version: number;
    parentDashboardId: string;
    dashboardName: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    tableOfContents?: any;
    description: string;
    createdBy: string;
    updatedAt: string;
    updatedBy?: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    deletedBy?: string;
    state: string;
    releaseNotes?: string;
    friendlyURL?: string;
}

// Public representation of a dashboard. Hides some fields
// like createdBy and state, which should not be exposed
// on public endpoints.
export interface PublicDashboard {
    id: string;
    name: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    tableOfContents?: any;
    description: string;
    updatedAt: Date;
    widgets?: Array<Widget>;
    friendlyURL?: string;

    // Filled on a homepage search with dashboard content
    // that matches the search query.
    queryMatches?: Array<string>;
}
