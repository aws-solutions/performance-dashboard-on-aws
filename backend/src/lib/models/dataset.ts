/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum SourceType {
    IngestApi = "IngestApi",
    FileUpload = "FileUpload",
}

export enum DatasetSchema {
    None = "None",
    Metrics = "Metrics",
}

export interface Dataset {
    id: string;
    createdBy: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    updatedAt: Date;
    sourceType: SourceType;
    schema?: DatasetSchema;
}

export type DatasetContent = Array<object>;
export type DatasetList = Array<Dataset>;

export interface DatasetItem {
    pk: string;
    sk: string;
    type: string;
    createdBy: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    updatedAt: string;
    sourceType: string;
    schema?: string;
}
