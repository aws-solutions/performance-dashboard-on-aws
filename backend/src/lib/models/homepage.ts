/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export interface Homepage {
    title: string;
    description: string;
    updatedAt: Date;
}

export interface HomepageItem {
    pk: string;
    sk: string;
    type: string;
    updatedAt?: string;
    title: string;
    description: string;
}
