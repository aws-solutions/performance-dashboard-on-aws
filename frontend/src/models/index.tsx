/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export type TopicArea = {
    id: string;
    name: string;
    createdBy: string;
    dashboardCount: number;
};

export const TopicAreaSortingCriteria = (a: TopicArea, b: TopicArea) => (a.name > b.name ? 1 : -1);

export type PublicTopicArea = {
    id: string;
    name: string;
    dashboards?: Array<Dashboard | PublicDashboard>;
};

export enum DashboardState {
    Draft = "Draft",
    Published = "Published",
    Archived = "Archived",
    PublishPending = "PublishPending",
    Inactive = "Inactive",
}

export enum SourceType {
    IngestApi = "IngestApi",
    FileUpload = "FileUpload",
}

export enum DatasetType {
    DynamicDataset = "DynamicDataset",
    StaticDataset = "StaticDataset",
    CsvFileUpload = "CsvFileUpload",
    CreateNew = "CreateNew",
}

export type Dashboard = {
    id: string;
    name: string;
    version: number;
    parentDashboardId: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description?: string;
    releaseNotes?: string;
    widgets: Array<Widget>;
    state: string;
    updatedAt: Date;
    createdBy: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    friendlyURL?: string;
    queryMatches?: Array<string>;
};

export type PublicDashboard = {
    id: string;
    name: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description?: string;
    widgets: Array<Widget>;
    updatedAt: Date;
    friendlyURL?: string;
    queryMatches?: Array<string>;
};

export type DashboardVersion = {
    id: string;
    version: number;
    state: DashboardState;
    friendlyURL?: string;
};

export type DashboardAuditLog = {
    itemId: string;
    timestamp: Date;
    version: number;
    userId: string;
    event: "Create" | "Update" | "Delete";
    modifiedProperties?: Array<{
        property: string;
        oldValue: any;
        newValue: any;
    }>;
};

export enum WidgetType {
    Text = "Text",
    Chart = "Chart",
    Table = "Table",
    Image = "Image",
    Metrics = "Metrics",
    Section = "Section",
}

export enum ChartType {
    LineChart = "LineChart",
    ColumnChart = "ColumnChart",
    BarChart = "BarChart",
    PartWholeChart = "PartWholeChart",
    PieChart = "PieChart",
    DonutChart = "DonutChart",
}

export interface Widget {
    id: string;
    name: string;
    widgetType: string;
    order: number;
    updatedAt: Date;
    dashboardId: string;
    content: any;
    showTitle: boolean;
    section?: string;
}

export interface ChartWidget extends Widget {
    content: {
        title: string;
        chartType: ChartType;
        datasetId: string;
        summary: string;
        summaryBelow: boolean;
        datasetType?: DatasetType;
        s3Key: {
            raw: string;
            json: string;
        };
        columnsMetadata: ColumnMetadata[];
        sortByColumn?: string;
        sortByDesc?: boolean;
        horizontalScroll?: boolean;
        stackedChart?: boolean;
        significantDigitLabels: boolean;
        dataLabels: boolean;
        computePercentages: boolean;
        showTotal: boolean;
    };
}

export interface TableWidget extends Widget {
    content: {
        title: string;
        datasetId: string;
        summary: string;
        summaryBelow: boolean;
        datasetType?: DatasetType;
        s3Key: {
            raw: string;
            json: string;
        };
        columnsMetadata: ColumnMetadata[];
        sortByColumn?: string;
        sortByDesc?: boolean;
        significantDigitLabels: boolean;
        displayWithPages: boolean;
    };
}

export interface ImageWidget extends Widget {
    content: {
        title: string;
        imageAltText: string;
        summary?: string;
        summaryBelow: boolean;
        s3Key: {
            raw: string;
        };
        filename: string;
        scalePct?: string;
    };
}

export interface MetricsWidget extends Widget {
    content: {
        title: string;
        datasetId: string;
        oneMetricPerRow: boolean;
        datasetType?: DatasetType;
        significantDigitLabels: boolean;
        metricsCenterAlign: boolean;
        s3Key: {
            raw: string;
            json: string;
        };
    };
}

export interface SectionWidget extends Widget {
    content: {
        title: string;
        summary: string;
        widgetIds?: Array<string>;
        showWithTabs: boolean;
        horizontally?: boolean;
    };
}

export enum DatasetSchema {
    None = "None",
    Metrics = "Metrics",
}

export type Dataset = {
    id: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    createdBy?: string;
    updatedAt?: Date;
    sourceType?: SourceType;
    schema?: DatasetSchema;
};

export type PublicHomepage = {
    title: string;
    description: string;
    dashboards: Array<PublicDashboard>;
};

export type Homepage = {
    title: string;
    description: string;
    updatedAt: Date;
};

export type PublicSettings = {
    dateTimeFormat: {
        date: string;
        time: string;
    };
    navbarTitle?: string;
    colors: {
        primary: string;
        secondary: string;
    };
    topicAreaLabels?: {
        singular: string;
        plural: string;
    };
    customLogoS3Key?: string;
    customFaviconS3Key?: string;
    contactEmailAddress?: string;
    contactUsContent?: string;
    analyticsTrackingId?: string;
};

export type Settings = {
    publishingGuidance: string;
    dateTimeFormat: {
        date: string;
        time: string;
    };
    updatedAt?: Date;
    navbarTitle?: string;
    colors: {
        primary: string;
        secondary: string;
    };
    topicAreaLabels: {
        singular: string;
        plural: string;
    };
    customLogoS3Key?: string;
    customLogoAltText?: string;
    customFaviconS3Key?: string;
    contactEmailAddress?: string;
    adminContactEmailAddress?: string;
    contactUsContent?: string;
    analyticsTrackingId?: string;
};

export type Metric = {
    title: string;
    value: number;
    changeOverTime?: string;
    percentage?: string;
    currency?: string;
    startDate?: string;
    endDate?: string;
};

// Type for the History object in react-router. Defines the
// location state that is common across all routes.
// Usage in a component: useHistory<LocationState>();
export type LocationState = {
    alert?: {
        type: "success" | "warning" | "info" | "error";
        message: string;
        to?: string;
        linkLabel?: string;
    };
    emails?: string;
    usernames?: Array<string>;
    id?: string;
    metrics?: Array<Metric>;
    metric?: Metric;
    position?: number;
    showTitle?: boolean;
    oneMetricPerRow?: boolean;
    significantDigitLabels?: boolean;
    metricsCenterAlign?: boolean;
    metricTitle?: string;
    origin?: string;
    json?: Array<any>;
    staticDataset?: Dataset;
    redirectUrl?: string;
    crumbLabel?: string;
    datasetType?: DatasetType;
};

export enum UserRoles {
    Admin = "Admin",
    Editor = "Editor",
    /**
     * @type {string}
     * @deprecated no longer supported
     */
    Publisher = "Publisher",
    Public = "Public",
}

export type User = {
    userId: string;
    enabled: boolean;
    userStatus: string;
    sub: string;
    email: string;
    roles: Array<UserRoles>;
    createdAt: Date;
    updatedAt: Date;
};

export enum ColumnDataType {
    Number = "Number",
    Text = "Text",
    Date = "Date",
}

export enum NumberDataType {
    Percentage = "Percentage",
    Currency = "Currency",
    "With thousands separators" = "With thousands separators",
}

export enum CurrencyDataType {
    "Dollar $" = "Dollar $",
    "Euro €" = "Euro €",
    "Pound £" = "Pound £",
}

export interface ColumnMetadata {
    columnName: string;
    dataType?: ColumnDataType;
    numberType?: NumberDataType;
    currencyType?: CurrencyDataType;
    hidden: boolean;
}
