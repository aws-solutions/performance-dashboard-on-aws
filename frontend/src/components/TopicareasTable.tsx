import React, { useState } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faChevronDown,
  faChevronCircleUp,
  faChevronCircleDown,
  IconDefinition,
} from "@fortawesome/free-solid-svg-icons";
import { TopicArea } from "../models";
import Button from "./Button";
import { useTranslation } from "react-i18next";

interface Props {
  topicAreas: Array<TopicArea>;
  onSelect?: Function;
}

interface SelectionHashMap {
  [topicareaId: string]: TopicArea;
}

type ColumnType = "name" | "createdBy" | "dashboards";
type Direction = "up" | "down";

function TopicareasTable(props: Props) {
  const { t } = useTranslation();
  const [selected, setSelected] = useState<SelectionHashMap>({});
  const [sortedBy, setSortedBy] = useState<ColumnType>("dashboards");
  const [direction, setDirection] = useState<Direction>("down");

  const onSelect = (topicarea: TopicArea) => {
    const selection = { [topicarea.id]: topicarea };
    setSelected(selection);
    onSelectCallback(selection);
  };

  const isSelected = (topicarea: TopicArea) => {
    return !!selected[topicarea.id];
  };

  const onSelectCallback = (selection: SelectionHashMap) => {
    if (props.onSelect) {
      props.onSelect(Object.values(selection));
    }
  };

  const sortBy = (columnType: ColumnType) => {
    if (sortedBy === columnType) {
      setDirection(direction === "down" ? "up" : "down");
    } else {
      setDirection("down");
      setSortedBy(columnType);
    }
  };

  const sortByIcon = (
    actualColunmName: string,
    expectedColumnName: string,
    direction: Direction
  ): IconDefinition => {
    if (actualColunmName !== expectedColumnName) {
      return faChevronDown;
    }
    return direction === "up" ? faChevronCircleUp : faChevronCircleDown;
  };

  return (
    <table className="usa-table usa-table--borderless" width="100%">
      <thead>
        <tr>
          <th></th>
          <th>
            <Button
              variant="unstyled"
              className="margin-left-1 hover:text-base-darker font-sans-md text-bold text-no-underline"
              onClick={() => sortBy("name")}
              ariaLabel={t("SortTopicAreaName")}
            >
              <span>{t("TopicArea")}</span>
              <FontAwesomeIcon
                className="margin-left-1"
                icon={sortByIcon(sortedBy, "name", direction)}
              />
            </Button>
          </th>
          <th>
            <Button
              variant="unstyled"
              className="margin-left-1 hover:text-base-darker font-sans-md text-bold text-no-underline"
              onClick={() => sortBy("dashboards")}
              ariaLabel={t("SortDashboards")}
            >
              <span>{t("Dashboards")}</span>
              <FontAwesomeIcon
                className="margin-left-1"
                icon={sortByIcon(sortedBy, "dashboards", direction)}
              />
            </Button>
          </th>
          <th>
            <Button
              variant="unstyled"
              className="margin-left-1 hover:text-base-darker font-sans-md text-bold text-no-underline"
              onClick={() => sortBy("createdBy")}
              ariaLabel={t("SortCreatedBy")}
            >
              <span>{t("CreatedBy")}</span>
              <FontAwesomeIcon
                className="margin-left-1"
                icon={sortByIcon(sortedBy, "createdBy", direction)}
              />
            </Button>
          </th>
        </tr>
      </thead>
      <tbody>
        {sortTopicAreas(props.topicAreas, sortedBy, direction).map(
          (topicarea) => (
            <tr key={topicarea.id}>
              <td>
                <label className="usa-sr-only" htmlFor={topicarea.id}>
                  {topicarea.name}
                </label>
                <input
                  type="radio"
                  id={topicarea.id}
                  checked={isSelected(topicarea)}
                  onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                    if (!isSelected(topicarea) && e.target.checked) {
                      onSelect(topicarea);
                    }
                  }}
                />
              </td>
              <td>
                <span className="text-bold text-base-darker font-sans-md">
                  {topicarea.name}
                </span>
              </td>
              <td>
                <span className="font-sans-md">
                  {topicarea.dashboardCount || "-"}
                </span>
              </td>
              <td>
                <span className="font-sans-md">{topicarea.createdBy}</span>
              </td>
            </tr>
          )
        )}
      </tbody>
    </table>
  );
}

function sortTopicAreas(
  topicAreas: Array<TopicArea>,
  sortedBy: ColumnType,
  direction: Direction
): Array<TopicArea> {
  const sortedTopicAreas = [...topicAreas];
  if (sortedBy === "name") {
    sortedTopicAreas.sort((a: TopicArea, b: TopicArea) =>
      direction === "down"
        ? a.name.localeCompare(b.name)
        : b.name.localeCompare(a.name)
    );
  } else if (sortedBy === "createdBy") {
    sortedTopicAreas.sort((a: TopicArea, b: TopicArea) =>
      direction === "down"
        ? a.createdBy.localeCompare(b.createdBy)
        : b.createdBy.localeCompare(a.createdBy)
    );
  } else if (sortedBy === "dashboards") {
    sortedTopicAreas.sort((a: TopicArea, b: TopicArea) =>
      direction === "down"
        ? b.dashboardCount - a.dashboardCount
        : a.dashboardCount - b.dashboardCount
    );
  }
  return sortedTopicAreas;
}

export default TopicareasTable;
