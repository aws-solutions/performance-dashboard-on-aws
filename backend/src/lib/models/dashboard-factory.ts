import { v4 as uuidv4 } from "uuid";
import { Dashboard, DashboardItem } from "./dashboard-models";
import { User } from "./user-models";
import topicareaFactory from "./topicarea-factory";

const DASHBOARD: string = 'Dashboard';

function createNew(name: string, topicAreaId: string, topicAreaName: string, user: User): Dashboard {
  return {
    id: uuidv4(),
    name,
    topicAreaId,
    topicAreaName,
    createdBy: user.userId,
  };
}

/**
 * Converts a Dashboard to a DynamoDB item.
 */
function toItem(dashboard: Dashboard): DashboardItem {
  let item: DashboardItem = {
    pk: itemId(dashboard.id),
    sk: itemId(dashboard.id),
    type: DASHBOARD,
    dashboardName: dashboard.name,
    topicAreaName: dashboard.topicAreaName,
    topicAreaId: topicareaFactory.itemId(dashboard.topicAreaId),
    description: dashboard.description,
    overview: dashboard.overview,
    createdBy: dashboard.createdBy,
  };
  return item;
}

/**
 * Converts a DynamoDB item into a Dashboard object
 */
function fromItem(item: DashboardItem): Dashboard {
  const id = item.pk.substring(10);
  let dashboard: Dashboard = {
    id,
    name: item.dashboardName,
    topicAreaId: item.topicAreaId.substring(10),
    topicAreaName: item.topicAreaName,
    description: item.description,
    overview: item.overview,
    createdBy: item.createdBy,
  }
  return dashboard;
}

function itemId(id: string): string { return `${DASHBOARD}#${id}` }

export default {
  toItem,
  createNew,
  fromItem,
  itemId,
};
