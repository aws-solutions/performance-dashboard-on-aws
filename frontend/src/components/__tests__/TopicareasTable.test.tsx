import React from "react";
import { render, fireEvent, act } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { TopicArea } from "../../models";
import TopicareasTable from "../TopicareasTable";

const topicArea: TopicArea = {
  id: "abc",
  name: "USWDS",
  createdBy: "johndoe",
};

test("renders an empty table", async () => {
  const wrapper = render(<TopicareasTable topicAreas={[]} />);
  expect(wrapper.container).toMatchSnapshot();
});

test("renders a table with topic areas", async () => {
  const topicAreas = [topicArea];
  const wrapper = render(<TopicareasTable topicAreas={topicAreas} />, {
    wrapper: MemoryRouter,
  });
  expect(wrapper.container).toMatchSnapshot();
});

test("onSelect function is called when user selects topicarea", async () => {
  const onSelect = jest.fn();
  const { getByLabelText } = render(
    <TopicareasTable topicAreas={[topicArea]} onSelect={onSelect} />,
    {
      wrapper: MemoryRouter,
    }
  );

  await act(async () => {
    const checkbox = getByLabelText("USWDS");
    fireEvent.click(checkbox);
  });

  expect(onSelect).toBeCalledWith(expect.arrayContaining([topicArea]));
});
