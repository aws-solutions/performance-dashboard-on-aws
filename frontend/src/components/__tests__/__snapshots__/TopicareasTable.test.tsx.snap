// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a table with topic areas 1`] = `
<div>
  <table
    class="usa-table usa-table--borderless"
    width="100%"
  >
    <thead>
      <tr>
        <th />
        <th>
          <span
            class="font-sans-xs"
          >
            Ministry
          </span>
          <button
            class="usa-button usa-button--unstyled margin-left-1 hover:text-base-light text-white"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down fa-w-14 "
              data-icon="chevron-down"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                fill="currentColor"
              />
            </svg>
          </button>
        </th>
        <th>
          <span
            class="font-sans-xs"
          >
            Dashboards
          </span>
          <button
            class="usa-button usa-button--unstyled margin-left-1 hover:text-base-light text-base-darkest"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down fa-w-14 "
              data-icon="chevron-down"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                fill="currentColor"
              />
            </svg>
          </button>
        </th>
        <th>
          <span
            class="font-sans-xs"
          >
            Created by
          </span>
          <button
            class="usa-button usa-button--unstyled margin-left-1 hover:text-base-light text-white"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down fa-w-14 "
              data-icon="chevron-down"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                fill="currentColor"
              />
            </svg>
          </button>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <label
            class="usa-sr-only"
            for="abc"
          >
            USWDS
          </label>
          <input
            id="abc"
            type="radio"
          />
        </td>
        <td>
          <span
            class="text-bold text-base-darkest font-sans-md"
          >
            USWDS
          </span>
        </td>
        <td>
          <span
            class="font-sans-md"
          >
            1
          </span>
        </td>
        <td>
          <span
            class="font-sans-md"
          >
            johndoe
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`renders an empty table 1`] = `
<div>
  <table
    class="usa-table usa-table--borderless"
    width="100%"
  >
    <thead>
      <tr>
        <th />
        <th>
          <span
            class="font-sans-xs"
          >
            Ministry
          </span>
          <button
            class="usa-button usa-button--unstyled margin-left-1 hover:text-base-light text-white"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down fa-w-14 "
              data-icon="chevron-down"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                fill="currentColor"
              />
            </svg>
          </button>
        </th>
        <th>
          <span
            class="font-sans-xs"
          >
            Dashboards
          </span>
          <button
            class="usa-button usa-button--unstyled margin-left-1 hover:text-base-light text-base-darkest"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down fa-w-14 "
              data-icon="chevron-down"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                fill="currentColor"
              />
            </svg>
          </button>
        </th>
        <th>
          <span
            class="font-sans-xs"
          >
            Created by
          </span>
          <button
            class="usa-button usa-button--unstyled margin-left-1 hover:text-base-light text-white"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down fa-w-14 "
              data-icon="chevron-down"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                fill="currentColor"
              />
            </svg>
          </button>
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
